Implement a capacity-allocated constructor for `DAGCircuit` in Rust.

raynelfss opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [x] I have added the tests to cover my changes.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
-->

### Summary
Resolves #13002
After #12550 merges, the following commits aim to provide a`with_capacity` alternative initializer for the `DAGCircuit` in Rust.

### Details and comments
**Will be rebased after #12550 merges**

These commits add the `DAGCircuit::with_capacity` constructor to the rust representation of `DAGCircuit`. This would allow us to allocate an approximation of the necessary space to successfully create a `DAGCircuit` with certain properties to avoid re-allocating space in memory.

This constructor takes the following arguments:
- `num_qubits` for the number of qubits in the specific circuit.
- `num_clbits` for the number of clbits in the circuit.
- `num_ops` for the number of operations in the circuit (Optional).
- `num_vars` for the number of variables in the circuit (Optional).

### Questions
- What is the best way of estimating the possible number of edges required for `num_ops`?
   - An operation (or a block) can have as many input/output qargs/cargs as possible
   - I used 3 (2 qubits + 1 classical or 3 qubits) as a baseline. But this is just speculation, if a better approach exists, let me in a comment.
 - Best way of estimating the names of operations in a circuit?
    - In a `DAGCircuit`, there can be 100 `op_nodes` of just the same operation or 100 different operations occurring just once.
    - I initialized it as `num_ops` but if a different opinion exists, let me know in a comment.

### Blockers
- [x] #12550

mtreinish commented:
https://github.com/Qiskit/qiskit/pull/12812 might be useful for you on this. This was me adding basically the same interface to the python space constructor for `DAGCircuit`. We get more flexibility with it in Rust including for things like the internal caches so I probably will rebase that PR on top of this after this merges (and add a python interface for the alternative constructor).

raynelfss committed:
Initial: implement fixed capacity constructor for DAGCircuit
- Implement `DAGCircuit` with `with_capacity` to create an initially allocated instance, for rust only.
- Implement `with_capacity` for `BitData` and `IndexInterner` that creates an initally allocated instance of each type.
- Other small tweaks and fixes.
- Add num_edges optional argument. If known, use `num_edges` argument to create a `DAGCircuit` with the exact number of edges.
- Leverage usage of new method in `copy_empty_like`.
- Use `..Default::default()` for `op_names` and `calibrations` fields in `DAGCircuit::with_capacity()`

raynelfss committed:
Fix: Adapt to #13033

raynelfss committed:
Merge branch 'main' into rusty-dag-with-capacity

mtreinish commented on crates/circuit/src/dag_circuit.rs:
```diff
@@ -1556,7 +1556,14 @@ def _format(operand):
+            py,
+            self.num_qubits(),
+            self.num_clbits(),
+            Some(self.num_vars()),
+            None,
```
Shouldn't this be:

```suggestion
            None,
            Some(self.num_vars()),
```

Otherwise we potentially will be overallocating if there are vars but no ops. Although it might make more sense to flip the arguments in the `with_capacity` signature as doing `qubits, clbits, vars, ops` makes more sense intuitively to me.

raynelfss replied:
I'll re-arrange the function signature for this as it makes more sense, thank you 🚀

mtreinish reviewed (commented):
Overall this LGTM it's a straightforward addition to the rust api for `DAGCircuit`. Just one inline comment where I think one argument is in the wrong position.

raynelfss committed:
Fix: Re-arrange the function arguments as per @mtreinish's review.
- The order now follows: qubits, clbits, vars, num_ops, edges. As per [Matthew's comment](https://github.com/Qiskit/qiskit/pull/12975#discussion_r1742426093).

mtreinish reviewed (approved):
LGTM, thanks for the quick update
